### PR TITLE
Improve Toy Heteroscedastic Datamodule and Tutorials

### PR DESCRIPTION
**Pull request by nilsleh**
This PR aims to improve the clarity of the `ToyHeteroscedasticDatamodule` that we use for the tutorials, to make it more rigorous and clear what the separate data splits represent and how we assess model performance based on those data splits.

### CODE REVIEW

**Comment by nilsleh** [comment]
@elcorto Just thought that it might be easier to discuss your input in a PR like this where it is easier to tag/comment code. Overall I agree, that much of the definitions/naming can be improved, so below are some thoughts.

* Normalization statistics are now "handwerklich sauberer" :) 
* we differentiated between a train, val, calib, and test set with the following idea/motivation
  * training set: the data available to train your model
  * validation set: data available to evaluate the model performance on unseen data (with iid assumption)
  * calibration set: I agree the lingo here is confusing, but from the ML literature side (that often neglects mathematical/statistical rigor) conformal prediction expects a data set that is separate from training and validation and is solely used to "calibrate" or fit the conformal prediction procedure (at least that is how we understood it from https://arxiv.org/pdf/2107.07511), and is separate from the "validation" set which is used to evaluate the fit of your underlying Quantile Regression model before "recalibrating" it with the calibration set to improve coverage according to the conformal prediction scenario
  * test set: completely unseen data that in a real case is never touched unless you are done with everything above. In a "real world" scenario it might be the case that this data has a different distribution than your training data
* With the above description, I would leave the names as is, except that there is a "new" `test` data variable which is IID data, and the current `test` variable should be renamed to `gt` which would yield overall 5 data splits/variables.

I am happy to fix all the tutorials to make this more clear and avoid confusion.

Thanks as always for your valuable input!

**Comment by elcorto** [comment]
As discussed offline, here are some additional comments and suggestions:

`ToyHeteroscedasticDatamodule`:

* the `sigma` argument is not used
* the `generate_y` callable requires the signature `f(x, noise: bool)`, that could be added to the doc string
* by default, have 3 splits train/val/test and have an argument such as `split_calib: bool` which, if `True`, generates a 4th split `calib` for conformal prediction
* `n_train` is the total number of points, not the train points after splitting, maybe rename (e.g. `n_points`)
* the split sizes are hard-coded, maybe provide arguments such as `test_fraction`, `val_fraction` and optionally `calib_fraction`
* the splitting as implemented at the moment would not produce fractions as indicated by the above arguments
* the z-score is calculated over all data, not only train and then applied to all other splits
* seeds are hard-coded

General:

* the function `polynomial_f3` could be renamed to `noisy_sine` or thereabouts
* only `polynomial_f3` fulfills the signature required (`noise` keyword)

Here is a draft implementation that fixes some of the above. It doesn't add the calib set, that's why I don't send a PR at this point. Please feel free to re-use parts of this to adjust `ToyHeteroscedasticDatamodule` in the way you see fit.


```py
import warnings
from sklearn.preprocessing import StandardScaler

def noisy_sine(x, noise=True):
    """(Noisy) sine function."""
    out = 7 * np.sin(x)
    if noise:
        if isinstance(x, np.ndarray):
            assert x.ndim == 1, "x is not an 1d array"
            nn = np.random.randn(len(x))
        else:
            nn = np.random.randn()
        out += 3 * np.abs(np.cos(x / 2)) * nn
    return out


# For backward compatibility
def polynomial_f3(*args, **kwds):
    """(Noisy) sine function."""
    warnings.warn("This function was renamed to noisy_sine().", DeprecationWarning)
    return noisy_sine(*args, **kwds)


class ToyHeteroscedasticDatamodule(LightningDataModule):
    """Implement Toy Dataset with heteroscedastic noise."""
    def __init__(
        self,
        x_min: int | float = -4,
        x_max: int | float = 4,
        n_points: int = 200,
        batch_size: int = 100,
        test_fraction: float = 0.2,
        val_fraction: float = 0.1,
        generate_y: Callable = noisy_sine,
        noise_seed: int = 42,
        split_seed: int = 42,
    ) -> None:
        """Define a heteroscedastic toy regression dataset.

        Inspired by (https://mapie.readthedocs.io/en/latest/examples_regression/
        1-quickstart/plot_heteroscedastic_1d_data.html#sphx-glr-examples-
        regression-1-quickstart-plot-heteroscedastic-1d-data-py)

        Split `n_points` data points into train, validation and test set. We
        provide the following arrays: X_<type>, Y_<type>, where <type> is one
        of:

        * all: train + val + test
        * train: all - test - val 
        * test: see `test_fraction`
        * val: validation, see `val_fraction`
        * gtext: noise-free ground truth on extended x-axis

        X and Y have shape (n_points_{type}, 1).

        Args:
            x_min: Minimum value of x range
            x_max: Maximum value of x range
            n_points : Number of train + test + validation points
            batch_size: batch size for data loader
            test_fraction: fraction of n_points for test set
            val_fraction: fraction of n_points for validation
                set
            generate_y: ground truth function with noise option, must have
                signature f(x, noise: bool)
            noise_seed: random seed for x points positions and y noise
            split_seed: random seed for train/test/val split
        """
        super().__init__()

        # TODO: use rng=np.random.default_rng(seed) and pass to noisy_sine()
        np.random.seed(noise_seed)

        self.batch_size = batch_size

        x = np.empty(n_points)
        y = np.empty(n_points)
        for k in range(n_points):
            rnd = np.random.rand()
            if rnd < 1 / 3.0:
                x[k] = np.random.normal(loc=x_min, scale=2.0 / 5.0)
            else:
                if rnd < 2.0 / 3.0:
                    x[k] = np.random.normal(loc=0.0, scale=0.9)
                else:
                    x[k] = np.random.normal(loc=x_max, scale=2.0 / 5.0)

        y = generate_y(x)

        # Same as x.unsqueeze(-1) in torch. For NN training, all arrays (input
        # and target) must have shape (n_samples, n_features). For the 1D
        # problem here, we have (n_samples, 1).
        self.X_all = x[:, None]
        self.Y_all = y[:, None]

        X_other, self.X_test, Y_other, self.Y_test = train_test_split(
            self.X_all, self.Y_all, test_size=test_fraction, random_state=split_seed
        )

        self.X_train, self.X_val, self.Y_train, self.Y_val = train_test_split(
            X_other,
            Y_other,
            test_size=val_fraction / (1 - test_fraction),
            random_state=split_seed,
        )

        # Ground truth for plotting (x,y) and prediction (x)
        xmin, xmax = self.X_all.min(), self.X_all.max()
        span = xmax - xmin
        self.X_gtext = np.linspace(
            xmin - span * 0.2, xmax + span * 0.2, int(n_points * 1.5)
        )[:, None]
        self.Y_gtext = generate_y(self.X_gtext, noise=False)

        # Fit scalers on train data
        scalers = dict(
            X=StandardScaler().fit(self.X_train), Y=StandardScaler().fit(self.Y_train)
        )

        # Apply scaling to all splits, convert to torch tensors
        for xy in ["X", "Y"]:
            for arr_type in ["train", "test", "val", "gtext", "all"]:
                arr_name = f"{xy}_{arr_type}"
                setattr(
                    self,
                    arr_name,
                    self._n2t(scalers[xy].transform(getattr(self, arr_name))),
                )

    @staticmethod
    def _n2t(x):
        return torch.from_numpy(x).type(torch.float32)

    def train_dataloader(self) -> DataLoader:
        ...
```

**Comment by elcorto** [comment]
As discussed offline, here are some additional comments and suggestions:

`ToyHeteroscedasticDatamodule`:

* the `sigma` argument is not used
* the `generate_y` callable requires the signature `f(x, noise: bool)`
* by default, have 3 splits train/val/test and have an argument such as `split_calib: bool` which, if `True`, generates a 4th split `calib` for conformal prediction
* `n_train` is the total number of points, not the train points after splitting, maybe rename (e.g. `n_points`)
* the split sizes are hard-coded, maybe provide arguments such as `test_fraction`, `val_fraction` and optionally `calib_fraction`
* the splitting as implemented at the moment would not produce fractions as indicated by the above arguments
* the z-score is calculated over all data, not only train and then applied to all other splits
* seeds are hard-coded

General:

* the function `polynomial_f3` could be renamed to `noisy_sine` or thereabouts
* only `polynomial_f3` fulfills the signature required (`noise` keyword)

Here is a draft implementation that fixes some of the above. It doesn't add the calib set, that's why I don't send a PR at this point. Please feel free to re-use parts of this to adjust `ToyHeteroscedasticDatamodule` in the way you see fit.


```py
import warnings
from sklearn.preprocessing import StandardScaler

def noisy_sine(x, noise=True):
    """(Noisy) sine function."""
    out = 7 * np.sin(x)
    if noise:
        if isinstance(x, np.ndarray):
            assert x.ndim == 1, "x is not an 1d array"
            nn = np.random.randn(len(x))
        else:
            nn = np.random.randn()
        out += 3 * np.abs(np.cos(x / 2)) * nn
    return out


# For backward compatibility
def polynomial_f3(*args, **kwds):
    """(Noisy) sine function."""
    warnings.warn("This function was renamed to noisy_sine().", DeprecationWarning)
    return noisy_sine(*args, **kwds)


class ToyHeteroscedasticDatamodule(LightningDataModule):
    """Implement Toy Dataset with heteroscedastic noise."""
    def __init__(
        self,
        x_min: int | float = -4,
        x_max: int | float = 4,
        n_points: int = 200,
        batch_size: int = 100,
        test_fraction: float = 0.2,
        val_fraction: float = 0.1,
        generate_y: Callable = noisy_sine,
        noise_seed: int = 42,
        split_seed: int = 42,
    ) -> None:
        """Define a heteroscedastic toy regression dataset.

        Inspired by (https://mapie.readthedocs.io/en/latest/examples_regression/
        1-quickstart/plot_heteroscedastic_1d_data.html#sphx-glr-examples-
        regression-1-quickstart-plot-heteroscedastic-1d-data-py)

        Split `n_points` data points into train, validation and test set. We
        provide the following arrays: X_<type>, Y_<type>, where <type> is one
        of:

        * all: train + val + test
        * train: all - test - val
        * test: see `test_fraction`
        * val: validation, see `val_fraction`
        * gtext: noise-free ground truth on extended x-axis

        X and Y have shape (n_points_{type}, 1).

        Args:
            x_min: Minimum value of x range
            x_max: Maximum value of x range
            n_points : Number of train + test + validation points
            batch_size: batch size for data loader
            test_fraction: fraction of n_points for test set
            val_fraction: fraction of n_points for validation
                set
            generate_y: ground truth function with noise option, must have
                signature f(x, noise: bool)
            noise_seed: random seed for x points positions and y noise
            split_seed: random seed for train/test/val split
        """
        super().__init__()

        # TODO: use rng=np.random.default_rng(seed) and pass to noisy_sine()
        np.random.seed(noise_seed)

        self.batch_size = batch_size

        x = np.empty(n_points)
        y = np.empty(n_points)
        for k in range(n_points):
            rnd = np.random.rand()
            if rnd < 1 / 3.0:
                x[k] = np.random.normal(loc=x_min, scale=2.0 / 5.0)
            else:
                if rnd < 2.0 / 3.0:
                    x[k] = np.random.normal(loc=0.0, scale=0.9)
                else:
                    x[k] = np.random.normal(loc=x_max, scale=2.0 / 5.0)

        y = generate_y(x)

        # Same as x.unsqueeze(-1) in torch. For NN training, all arrays (input
        # and target) must have shape (n_samples, n_features). For the 1D
        # problem here, we have (n_samples, 1).
        self.X_all = x[:, None]
        self.Y_all = y[:, None]

        X_other, self.X_test, Y_other, self.Y_test = train_test_split(
            self.X_all, self.Y_all, test_size=test_fraction, random_state=split_seed
        )

        self.X_train, self.X_val, self.Y_train, self.Y_val = train_test_split(
            X_other,
            Y_other,
            test_size=val_fraction / (1 - test_fraction),
            random_state=split_seed,
        )

        # Ground truth for plotting (x,y) and prediction (x)
        xmin, xmax = self.X_all.min(), self.X_all.max()
        span = xmax - xmin
        self.X_gtext = np.linspace(
            xmin - span * 0.2, xmax + span * 0.2, int(n_points * 1.5)
        )[:, None]
        self.Y_gtext = generate_y(self.X_gtext, noise=False)

        # Fit scalers on train data
        scalers = dict(
            X=StandardScaler().fit(self.X_train), Y=StandardScaler().fit(self.Y_train)
        )

        # Apply scaling to all splits, convert to torch tensors
        for xy in ["X", "Y"]:
            for arr_type in ["train", "test", "val", "gtext", "all"]:
                arr_name = f"{xy}_{arr_type}"
                setattr(
                    self,
                    arr_name,
                    self._n2t(scalers[xy].transform(getattr(self, arr_name))),
                )

    @staticmethod
    def _n2t(x):
        return torch.from_numpy(x).type(torch.float32)

    def train_dataloader(self) -> DataLoader:
        ...
```

**Comment by nilsleh** [comment]
> As discussed offline, here are some additional comments and suggestions:

Thanks for all the comments, will push a version to incorporate all the suggestions, with your cleaner code, in the next few days.

**Comment by elcorto** [comment]
Thanks & no rush :) Let me know if you need a review then.

**Comment by nilsleh** [comment]
While working on this, I now actually realize, what the motivation was for having "X_test" be the dense noise free "ground truth". And it was mainly so the plots would look better :smile:  if you don't plot over the whole X-axis, and just plot the predictions over the IID samples the plots don't look as nice, because we have gapped data and also just a fixed number of samples (and I don't want to increase the number of samples that much, because then all the notebooks take longer).

The new IID X_test:
Screenshot from 2024-05-30 19-42-06

The X_gtext:

Screenshot from 2024-05-30 19-41-36

So I think for the plot, evaluating over the whole line so `X_gtext` makes sense, but then we want to make a second prediction on `X_test` for the calibration plot?

**Comment by elcorto** [comment]
Yes that would be the idea, `X_gtext` for plotting predictions and confidence band, and `X_test` for calibration.

File: docs/tutorials/regression/bnn_vi_elbo.ipynb [hunk 1]
```diff
@@ -329,30 +331,14 @@
...
-  },
   {
    "cell_type": "markdown",
    "id": "f9b45c30",
    "metadata": {},
    "source": [
-    "## Evaluate Predictions"
+    "## Evaluate Predictions\n",
+    "\n",
+    "The constructed Data Module contains two possible test variable. `X_test` are IID samples from the same noise distribution as the training data, while `X_gtext` are dense inputs from the underlying \"ground truth\" function without any noise that also extends the input range to either side, so we can visualize the method's UQ tendencies when extrapolating beyond the training data range. Thus, we will use `X_gtext` for visualization purposes, but use `X_test` to compute uncertainty and calibration metrics because we want to analyse how well the method has learned the noisy data distribution."
```
Comment (nilsleh) [inline docs/tutorials/regression/bnn_vi_elbo.ipynb:341]: @elcorto This is a draft for the proposed text, I would add to the notebooks to explain what variables we use for what and why. If we settle on one, I will add it to the other notebooks.

**Comment by nilsleh** [comment]
Closes #155 as well.
